src/item/enchantment/AvailableEnchantmentRegistry.php

```diff
@@ -0,0 +1,183 @@
+		);
+		$this->register(Enchantments::VANISHING(), [], [Tags::ALL]);
+		$this->register(Enchantments::SWIFT_SNEAK(), [], [Tags::LEGGINGS]);
+	}
+
+	/**
+	 * @param string[] $primaryItemTags
+	 * @param string[] $secondaryItemTags
+	 */
+	public function register(Enchantment $enchantment, array $primaryItemTags, array $secondaryItemTags) : void{
+		$this->enchantments[spl_object_id($enchantment)] = $enchantment;
+		$this->setPrimaryItemTags($enchantment, $primaryItemTags);
+		$this->setSecondaryItemTags($enchantment, $secondaryItemTags);
+	}
+
+	public function unregister(Enchantment $enchantment) : void{
+		unset($this->enchantments[spl_object_id($enchantment)]);
+		unset($this->primaryItemTags[spl_object_id($enchantment)]);
+		unset($this->secondaryItemTags[spl_object_id($enchantment)]);
+	}
+
+	public function unregisterAll() : void{
+		$this->enchantments = [];
+		$this->primaryItemTags = [];
+		$this->secondaryItemTags = [];
+	}
+
+	public function isRegistered(Enchantment $enchantment) : bool{
+		return isset($this->enchantments[spl_object_id($enchantment)]);
+	}
+
+	/**
+	 * Returns item tags of the specified enchantment that are used to determine the available enchantments for the
+	 * item for any type of in-game enchanting: enchanting table, anvil, fishing, etc.
+	 *
+	 * @return string[]
+	 */
+	public function getPrimaryItemTags(Enchantment $enchantment) : array{
+		return $this->primaryItemTags[spl_object_id($enchantment)] ?? [];
+	}
+
+	/**
+	 * @param string[] $tags
+	 */
+	public function setPrimaryItemTags(Enchantment $enchantment, array $tags) : void{
+		if(!$this->isRegistered($enchantment)){
+			throw new \LogicException("Cannot set primary item tags for non-registered enchantment");
+		}
+		$this->primaryItemTags[spl_object_id($enchantment)] = array_values($tags);
+	}
+
+	/**
+	 * Returns item tags of the specified enchantment that are not used for an enchanting table, but are used for other
+	 * types of in-game enchanting: anvil, fishing, etc.
+	 *
+	 * @return string[]
+	 */
+	public function getSecondaryItemTags(Enchantment $enchantment) : array{
+		return $this->secondaryItemTags[spl_object_id($enchantment)] ?? [];
+	}
+
+	/**
+	 * @param string[] $tags
+	 */
+	public function setSecondaryItemTags(Enchantment $enchantment, array $tags) : void{
+		if(!$this->isRegistered($enchantment)){
+			throw new \LogicException("Cannot set secondary item tags for non-registered enchantment");
+		}
+		$this->secondaryItemTags[spl_object_id($enchantment)] = array_values($tags);
+	}
+
+	/**
+	 * Returns enchantments that can be applied to the specified item.
+	 *
+	 * @param bool $onlyPrimaryTags Whether to check only primary item tags for each enchantment.
+	 * This should only be used to select available enchantments for an enchanting table.
+	 *
+	 * @return Enchantment[]
+	 */
+	public function getAvailableEnchantments(Item $item, bool $onlyPrimaryTags = false) : array{
```

Comment:
Prefer making two functions for this. Boolean parameters are opaque and hard to read in code. Also, don't use optional parameters.